### ISSUE TITLE
Could not reach Cloud Firestore backend.

### ISSUE THREAD

**Issue by h2u** [firebase/firebase-js-sdk#1190]
### Environment

  * Operating System version: Windows 10
  * Browser version: Chrome 68.0.3440.106
  * Firebase SDK version: 5.4.2
  * Firebase Product: store

### Problem

Im using Firestore with Ionic (Angularfire2 package). Any collection or doc gives error but after ~10 seconds it gives data. Auth works without any problems.

Error:

```
@firebase/firestore: Firestore (5.0.4): Could not reach Cloud Firestore backend. Backend didn't respond within 10 seconds.
This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.
```

Created new Firebase Project and new Ionic project, installed packages, tried to make collection, shows same error.

```
    "angularfire2": "^5.0.0-rc.12",
    "firebase": "^5.4.2",
    "rxjs": "^6.3.1",
    "rxjs-compat": "^6.3.1",
```

The code im trying to run:

```
      db.collection('categories').valueChanges().subscribe(data => {
        console.log(data);
      });
```

And the console output:

```
[2018-09-02T09:05:36.388Z]  @firebase/firestore: Firestore (5.0.4) [FirestoreClient]: Initializing. user= null
index.esm.js:65 [2018-09-02T09:05:36.389Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Start LocalStore
index.esm.js:65 [2018-09-02T09:05:36.391Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Get last stream token
index.esm.js:65 [2018-09-02T09:05:36.392Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Get next mutation batch
index.esm.js:65 [2018-09-02T09:05:36.394Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Allocate query
index.esm.js:65 [2018-09-02T09:05:36.396Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Execute query
index.esm.js:65 [2018-09-02T09:05:36.396Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Remote document keys
index.esm.js:65 [2018-09-02T09:05:36.399Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: Creating WebChannel: https://firestore.googleapis.com/google.firestore.v1beta1.Firestore/Listen/channel [object Object]
index.esm.js:65 [2018-09-02T09:05:36.415Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: Opening WebChannel transport.
index.esm.js:65 [2018-09-02T09:05:36.418Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel sending: {"database":"projects/app-17s82/databases/(default)","addTarget":{"query":{"structuredQuery":{"from":[{"collectionId":"categories"}],"orderBy":[{"field":{"fieldPath":"__name__"},"direction":"ASCENDING"}]},"parent":"projects/app-17s82/databases/(default)"},"targetId":2}}
index.esm.js:65 [2018-09-02T09:05:36.683Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel transport opened.
index.esm.js:77 [2018-09-02T09:05:47.151Z]  @firebase/firestore: Firestore (5.0.4): Could not reach Cloud Firestore backend. Backend didn't respond within 10 seconds. This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.
index.esm.js:65 [2018-09-02T09:06:06.989Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"targetChange":{"targetChangeType":"ADD","targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:06.991Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:32.276692Z","fields":{"image":{"stringValue":"categories/a.jpg"},"lang_key":{"stringValue":"A"},"order":{"integerValue":"10"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/54Nxy6OOm4yantdzUDXu","updateTime":"2018-08-27T18:08:13.536914Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:06.994Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:31.537482Z","fields":{"image":{"stringValue":"categories/b.jpg"},"lang_key":{"stringValue":"B"},"order":{"integerValue":"6"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/8xNLqdlqqflMbu6Y8P8k","updateTime":"2018-08-26T18:59:31.537482Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:06.996Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:30.510370Z","fields":{"image":{"stringValue":"categories/c.jpg"},"lang_key":{"stringValue":"C"},"order":{"integerValue":"1"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/9UFJB2KAhFij6tcKaVFK","updateTime":"2018-08-26T18:59:30.510370Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:06.997Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:31.944855Z","fields":{"image":{"stringValue":"categories/d.jpg"},"lang_key":{"stringValue":"D"},"order":{"integerValue":"8"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/B76lIXguh3hFZcnXLybt","updateTime":"2018-08-26T18:59:31.944855Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:06.999Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:31.144277Z","fields":{"image":{"stringValue":"categories/e.jpg"},"lang_key":{"stringValue":"E"},"order":{"integerValue":"4"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/Dh10cOAFtywBF1nD6Oit","updateTime":"2018-08-26T18:59:31.144277Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.000Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:32.469426Z","fields":{"image":{"stringValue":"categories/a1.jpg"},"lang_key":{"stringValue":"A1"},"order":{"integerValue":"11"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/TaMikooiiCQSVcgvdUcM","updateTime":"2018-08-26T18:59:32.469426Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.002Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:31.341506Z","fields":{"image":{"stringValue":"categories/a2.jpg"},"lang_key":{"stringValue":"A2"},"order":{"integerValue":"5"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/bNk3WF02m9QC72W9CNFD","updateTime":"2018-08-26T18:59:31.341506Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.003Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:32.137302Z","fields":{"image":{"stringValue":"categories/a3.jpg"},"lang_key":{"stringValue":"A3"},"order":{"integerValue":"9"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/e0ffuG7GM1ikxdksjihZ","updateTime":"2018-08-26T18:59:32.137302Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.004Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:30.946686Z","fields":{"image":{"stringValue":"categories/a4.jpg"},"lang_key":{"stringValue":"A4"},"order":{"integerValue":"3"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/fgxGcgA9Ye3sue2FkNzo","updateTime":"2018-08-26T18:59:30.946686Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.006Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:31.747379Z","fields":{"image":{"stringValue":"categories/a5.jpg"},"lang_key":{"stringValue":"A5"},"order":{"integerValue":"7"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/fnSThPm2vbDdO8w5fIA1","updateTime":"2018-08-26T18:59:31.747379Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.008Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:30.294481Z","fields":{"image":{"stringValue":"categories/b1.jpg"},"lang_key":{"stringValue":"WORK"},"B1":{"integerValue":"0"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/tDQRayJlugtkIvIdtqGb","updateTime":"2018-08-26T18:59:30.294481Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.010Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"documentChange":{"document":{"createTime":"2018-08-26T18:59:30.712287Z","fields":{"image":{"stringValue":"categories/b2.jpg"},"lang_key":{"stringValue":"B2"},"order":{"integerValue":"2"},"visible":{"booleanValue":true}},"name":"projects/app-17s82/databases/(default)/documents/categories/tymqkjiII9sEKWxR9Cnw","updateTime":"2018-08-26T18:59:30.712287Z"},"targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.011Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"targetChange":{"readTime":"2018-09-02T10:06:51.233599Z","resumeToken":"CgkIv6qFq4mc3QI=","targetChangeType":"CURRENT","targetIds":[2]}}
index.esm.js:65 [2018-09-02T09:06:07.011Z]  @firebase/firestore: Firestore (5.0.4) [Connection]: WebChannel received: {"targetChange":{"readTime":"2018-09-02T10:06:51.233599Z","resumeToken":"CgkIv6qFq4mc3QI="}}
index.esm.js:65 [2018-09-02T09:06:07.022Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Get last remote snapshot version
index.esm.js:65 [2018-09-02T09:06:07.026Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Apply remote event
index.esm.js:65 [2018-09-02T09:06:07.043Z]  @firebase/firestore: Firestore (5.0.4) [MemoryPersistence]: Starting transaction: Garbage collection
app.provider.ts:62 (12) [{…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}, {…}]
```

**Comment by nguyenviettung7691** [comment]
Encountered the same error just yesterday when querying data. Seems to be a problem with Firestore not able to get data.

**Environment**
Operating System version: Windows 10
Browser version: Chrome 68.0.3440.106
Firebase SDK version: 5.0.4
Firebase Product: store

**Packages**
```
"dependencies": {
    "animated-number-vue": "^0.1.2",
    "firebase": "^5.0.4",
    "highcharts": "^6.1.0",
    "mdi": "^2.2.43",
    "raven-js": "^3.26.2",
    "vue": "^2.5.16",
    "vue-add-to-calendar": "^1.0.4",
    "vue-fullscreen": "^2.1.3",
    "vue-highcharts": "0.0.10",
    "vue-infinite-scroll": "^2.0.2",
    "vue-progressbar": "^0.7.5",
    "vue-router": "^3.0.1",
    "vue-swatches": "^1.0.1",
    "vuefire": "^1.4.5",
    "vuetify": "^1.0.17",
    "xlsx": "^0.13.0"
  },
  "devDependencies": {
    "@babel/core": "^7.0.0-beta.51",
    "@babel/plugin-syntax-jsx": "^7.0.0-beta.51",
    "@babel/plugin-transform-runtime": "^7.0.0-beta.51",
    "@babel/preset-env": "^7.0.0-beta.51",
    "@babel/preset-stage-2": "^7.0.0-beta.51",
    "@vue/cli": "^3.0.0-rc.2",
    "@vue/cli-init": "^3.0.0-rc.2",
    "autoprefixer": "^8.3.0",
    "babel-helper-vue-jsx-merge-props": "^2.0.3",
    "babel-loader": "^8.0.0-beta.3",
    "babel-plugin-transform-vue-jsx": "^4.0.1",
    "chalk": "^2.4.1",
    "copy-webpack-plugin": "^4.5.1",
    "css-loader": "^0.28.11",
    "file-loader": "^1.1.11",
    "friendly-errors-webpack-plugin": "^1.7.0",
    "html-webpack-plugin": "^3.2.0",
    "less": "^3.0.4",
    "less-loader": "^4.1.0",
    "mini-css-extract-plugin": "^0.4.0",
    "node-less": "^1.0.0",
    "node-notifier": "^5.2.1",
    "optimize-css-assets-webpack-plugin": "^4.0.0",
    "ora": "^2.0.0",
    "portfinder": "^1.0.13",
    "postcss-import": "^11.1.0",
    "postcss-loader": "^2.1.5",
    "postcss-url": "^7.2.1",
    "rimraf": "^2.6.2",
    "semver": "^5.5.0",
    "shelljs": "^0.8.1",
    "uglifyjs-webpack-plugin": "^1.1.1",
    "url-loader": "^1.0.1",
    "vue-loader": "^15.0.3",
    "vue-style-loader": "^4.1.0",
    "vue-template-compiler": "^2.5.16",
    "webpack": "^4.6.0",
    "webpack-bundle-analyzer": "^2.13.1",
    "webpack-cli": "^3.0.7",
    "webpack-dev-server": "^3.1.3",
    "webpack-merge": "^4.1.3"
  },
```

**Comment by banazak** [comment]
+1

**Comment by d-yoshi** [comment]
I think that anti-virus software is causing this error as one of the causes.

**Comment by mickev36** [comment]
@h2u Is the problem random or occuring for each request ?
We have a similar issue here for a similar setup (firestore 5.0.4, chrome), and this problem occurs randomly, blocking all requests after the first fail. Problem is happening in production for our users.

I managed to find out that installing the Firebug Lite Chrome extension makes the problem appear almost immediately and consistently. The "XHR Listener" used by firebug (patching the XMLHttpRequest with a wrapper) is causing the issue. Unfortunately, out users/clients encounter the problem without Firebug Lite installed, so the root cause must be somewhere else.

Any help would be appreciated.

**Comment by h2u** [comment]
@mickev36 , the problem is random but happens very recently.

Maybe @d-yoshi is right, it's anti-virus. Had no chance to work on project without anti-virus on Windows. 

At the moment Im working on MacOS on the same network as Windows was and have no problems with connection.

**Comment by banazak** [comment]
In our case it was also the anti-virus (Kaspersky).  It worked without it.

**Comment by h2u** [comment]
I feel that I have same problem because I have Kaspersky as well.

**Comment by d-yoshi** [comment]
I'm also using Kaspersky and I encountered this problem frequently.
I don't know any solution other than invalidating Kaspersky for now. 😢

**Comment by nguyenviettung7691** [comment]
I can confirm that this bug is very random, happened recently, and I don't have any antivirus software, except the default Windows Defender, which I can't turn off because of company policy.

**Comment by gamsterx** [comment]
Same problem here since a week or two. Also Kaspersky... 
When I first load the page the following error message is output (every single time):

> @firebase/firestore: Firestore (5.5.0): Could not reach Cloud Firestore backend. Backend didn't respond within 10 seconds.

after 30+ seconds the data from firestore finally loads 

>XHR finished loading: POST "https://firestore.googleapis.com/google.firestore.v1beta1.Firestore/Listen/channel..."

If I reload the page and navigate to another page, where firestore data is loaded, I immediately get the data back without the error.  😕 

If I pause Kaspersky the error does not occurr. 

```
...
"@angular/fire": "^5.0.0",
"firebase": "^5.5.0",
...
```

**Comment by ThiloKielmann** [comment]
Same here. My users (in web browsers) get the same message about firestore not being able to reach its backend once Kaspersky is activated on a Win 10 machine...

**Comment by Geoffrey-Pliez** [comment]
Same for me but only on IE 11.

But if I load the page, navigate immediatly to another page and then return to "firestore page", data is loaded without the error.

**Comment by chunkydonut21** [comment]
In my case it was due to Rn-fetch-blob library.
I moved ```
const Blob = RNFetchBlob.polyfill.Blob
const fs = RNFetchBlob.fs
window.XMLHttpRequest = RNFetchBlob.polyfill.XMLHttpRequest
window.Blob = Blob
``` inside the async function rather then defining it at the top. And it solved.

**Comment by rodrigomata** [comment]
+1

**Comment by gamsterx** [comment]
So I just generated a new angular project and only used the firebase sdk (5.5.2) to see if some other package causes the issue. It had the same outcome. 

I currently have Kaspersky Internet Security 2018 installed. When I switch off the web anti virus feature the error does not occur! I immediately get all my desired data back from firestore.

My current workaround for development is to leave web anti virus running and to add the following url to the tursted websites within the web anti virus settings of kaspersky: 

` https://firestore.googleapis.com/*`

This seems to work, but would only be acceptable for development.

**Comment by ddesna** [comment]
I'm working on mac os (without any Kaspersky 😆) and I have the same problem:
```
@firebase/firestore: Firestore (5.5.0): Could not reach Cloud Firestore backend. Connection failed 1 times. Most recent error: FirebaseError: [code=failed-precondition]: The Cloud Firestore API is not available for Cloud Datastore projects.
This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.
```

**Comment by wilhuff** [comment]
That's a totally different problem:

Most recent error: FirebaseError: [code=failed-precondition]: The Cloud Firestore API is not available for Cloud Datastore projects.

You're trying to use the Firestore API against a project that was already enabled for Cloud Datastore.

**Comment by iffj** [comment]
same here. In my case it was due to the Adguard I was using

**Comment by mikelehen** [comment]
Thanks @pitw !  Much appreciated.

**Comment by steelcorsy** [comment]
> Have same Problem with Kaspersky + Firestore.
> @mikelehen @gamsterx Report to Kaspersky is done, hopefully they lower the restrictions.
> https://forum.kaspersky.com/index.php?/topic/402132-kaspersky-blocks-firestore/

Response of Kaspersky:
According with some posts included on the link : The problem described happen

randomly and is Not only caused by kaspersky, but for other sources too..

So better contact "github.com" or Kaspersky Technical support

at this weblink :  https://my.kaspersky.com/support

Then?

**Comment by pitw** [comment]
@mikelehen Its reported. We have now a trackable issue from Kasperksy. Will report here what they think.

**Comment by arturcarvalho** [comment]
I've been getting several exceptions caught by Sentry. I've got exceptions with Windows 10 and iOS (image below). I'm getting a lot on Windows 10.

I assume this is the same issue although I don't know how to replicate it. 

Is there anything I can do to debug this further? Does reloading the page solve the problem?

Thanks.

![timeoutbug](https://user-images.githubusercontent.com/2487710/47104828-9a1bfc80-d242-11e8-83a4-d2609e33ec86.jpg)

**Comment by mikelehen** [comment]
@arturcarvalho That error message will show up when offline or on a flaky/slow internet connection, so some percentage of your customers will hit it on a regular basis.  It mostly exists to help you (the developer) during debugging. 

That said, if you're getting customer reports of your app not working, that is a different matter.

**Comment by arturcarvalho** [comment]
Thanks @mikelehen. I'll put some kind of reload screen on the site and see how it goes.

**Comment by mikelehen** [comment]
I've repurposed this issue to specifically track the issues with Firestore in conjunction with Kaspersky antivirus.  For folks having other issues, please open a separate issue.

For those encountering the issue with Kaspersky, I'd appreciate it if one or two of you could test out a custom build to help us better understand the problem. Unfortunately even if this build works, it would come at the expense of performance and so it's likely not something we can ship (at least on by default).  To try it:
* If you're loading firestore via a <script include=...> then can you point it at: https://mike-shared.firebaseapp.com/firebase-firestore.js
* If you're loading firestore via npm, please install the custom build via: npm install https://mike-shared.firebaseapp.com/firebase-firestore-0.8.3.tgz

Please make sure you see a log message in the console saying "THIS IS A CUSTOM FIRESTORE BUILD FOR DEBUGGING CONNECTION ISSUES" which will confirm you're picking up the right build.

If it works as-is, let me know.  Else if it does not work, can you try this (and send me the output):

```
const db = firebase.firestore();
db.settings({
  timestampsInSnapshots: true,
  offlineTimeout: 75  // default is 10 seconds.
});

// Also enable debug logging:
firebase.firestore.setLogLevel('debug')
```

Additionally, I'm curious to confirm if Firebase Realtime Database is affected by Kaspersky or not.  So could folks try visiting https://www.firebase.com/test.html and send the output (perhaps as a gist)?

Thanks!

**Comment by mikelehen** [comment]
@seanw Unfortunately the symptom that triggers the log message (our connection to the backend not succeeding in a timely manner) is completely expected in many cases (i.e. you're intentionally offline or on a flakey network connection). So we can't make it an error.

As for behavior of our APIs though, you can essentially detect this.  Just create a listener with `{ includeMetadataChanges: true }` option and then do a listen.  If you get an event with `snapshot.metadata.fromCache === true` and don't get a subsequent event with `snapshot.metadat.fromCache === false` in a timely manner (use some sort of timeout), you can assume we're not able to reach the backend and you can show some sort of message to the user.

Hope this helps.

**Comment by joshorvis** [comment]
So I was having this same issue and was subscribed to this thread hoping for a solution - but then my father in law (a security guy) told me that using Kaspersky is a horrible idea.  So I now suggest that the solution here is to uninstall Kaspersky wherever possible!

https://www.wired.com/story/kaspersky-russia-antivirus/

https://www.theregister.co.uk/2018/05/15/kaspersky_labs_announces_move_to_zurich_dutch_government_questions_firm/

https://www.bbc.com/news/uk-42202191

**Comment by JonasJW** [comment]
Yep got the same problem on a Computer with Kaspersky. Turning it off will work but I'm worried about casual users that won't know its because Kaspersky and assume the site is broken...

**Comment by mikelehen** [comment]
For folks running into this issue, I'd be curious if somebody could try using this custom build:

Install via npm:
```
npm install https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-v5.7.0.tgz https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-firestore-v0.9.0.tgz https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-webchannel-wrapper-v0.2.11.tgz
```

Or you can point your `<script include=...>` at: https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-firestore.js

And then initialize the SDK with a new forceLongPolling option:

```
  const db = firebase.firestore();
  db.settings({forceLongPolling: true});
```

You should see a log message saying:

> WEBCHANNEL: ForceLongPolling enabled.

And then I'll be curious to know if your connection issue goes away. Please also provide as much detail as you can about your environment (browser, antivirus software / version, network details, etc.).  Thanks!

**Comment by pitw** [comment]
In the beginning we had this issue with angular + firestore.
Now switched to Vue.js but still have same error.

I still believe that Problem is on side of Kaspersky - tried many times to get some useful answer but except some "default answers" we didn't get any further.

It doesnt seem to concern all versions of Kaspersky (Endpoint Protection seems to work, Internet Security 2017 - 2019 seems to have this bug)

In case sb is interested we also opened up a thread on Kaspersky Forum here:
[https://forum.kaspersky.com/index.php?/topic/402132-kaspersky-blocks-firestore/](https://forum.kaspersky.com/index.php?/topic/402132-kaspersky-blocks-firestore/
)
Feel free to post some experiences there ;)

**Comment by sydfzlbshr** [comment]
For those who are still facing this issue and want to simulate the scenario without Kaspersky Antivirus, you can easily do it with Chrome's Dev Tools. Open the **Network** tab and reload the page. When a connection is attempted to firestore you'll see a request made with a name starting with **channel?database= ... **. Right click on that request and click on **Block request domain**.

Sadly I couldn't find a way to simulate it in other browsers. Please share if you do.

**Comment by sydfzlbshr** [comment]
> For folks running into this issue, I'd be curious if somebody could try using this custom build:
> 
> Install via npm:
> 
> ```
> npm install https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-v5.7.0.tgz \
>   https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-firestore-v0.9.0.tgz \
>   https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-webchannel-wrapper-v0.2.11.tgz
> ```
> Or you can point your `<script include=...>` at: https://mike-shared.firebaseapp.com/webchannel-proxy-issue/firebase-firestore.js
> 
> And then initialize the SDK with a new forceLongPolling option:
> 
> ```
>   const db = firebase.firestore();
>   db.settings({forceLongPolling: true});
> ```
> You should see a log message saying:
> 
> > WEBCHANNEL: ForceLongPolling enabled.
> 
> And then I'll be curious to know if your connection issue goes away. Please also provide as much detail as you can about your environment (browser, antivirus software / version, network details, etc.). Thanks!

@mikelehen Will ForceLongPolling solve the problem of flaky internet connections? I'm unable to simulate such a connection but my users seem to be having such issues.

**Comment by mikelehen** [comment]
@sydfzlbshr I'm not sure what issue you're referring to.  You are welcome to try forceLongPolling and see if it has an effect, but if you're dealing with an issue NOT related to Kaspersky, please open a new github issue. This one is specifically for Kaspersky.

Also, blocking Firestore in chrome dev tools is going to completely break Firestore and make it  impossible to work.  I don't think we have evidence to say that this is what Kaspersky is doing.  So I don't think it's an equivalent test.

**Comment by sydfzlbshr** [comment]
@mikelehen 

I installed the custom build and added the options as specified.

`    firebase.initializeApp(config);`
`    this.fireStoreDB = firebase.firestore();`
`    this.fireStoreDB.settings({ forceLongPolling: true });`
`    console.log('settings', this.fireStoreDB)`

the above console log contains the settings as:
![firestoresettings](https://user-images.githubusercontent.com/27231393/51531022-513b9b80-1e62-11e9-9ef2-29ce22652fd7.png)

But, i'm not getting the log message you had mentioned
> WEBCHANNEL: ForceLongPolling enabled.

Also, still unable to reach firestore backend.

Am I missing something?

**Comment by mikelehen** [comment]
I have a contact within Kaspersky asking for additional tracing, screenshots, etc. to help track this down.  Is there somebody that could help out? @pitw maybe?  Shoot me an email at michael@firebase.com.  Thanks!

**Comment by gamsterx** [comment]
@mikelehen What kind of screenshots do you need? I am able to reproduce the issue.

**Comment by mikelehen** [comment]
@gamsterx Thanks.  Actually, I think I've been able to help them reproduce it so I may not need any more info for now.  Will update if/when I learn more.  Thanks!

**Comment by sydfzlbshr** [comment]
@mikelehen I've tried your custom build and it works! :100:

MacOS Mojave, Safari, Kaspersky Internet Security

When will this be included in the official release?

**Comment by mikelehen** [comment]
@sydfzlbshr Thanks for the confirmation!  I'm looking into adding the `forceLongPolling` flag to an official release, but it will probably take a bit of time.  In addition, we don't intend to enable it by default because it causes additional latency and bandwidth usage, so you would need to manually opt-in to the behavior.

So I'm still hoping Kaspersky can address this on their end.

**Comment by mikelehen** [comment]
I've received word from Kasparov that they have a specific plan for addressing this which should be released in the next 2-4 weeks.

**Comment by Ferrasil** [comment]
> > WEBCHANNEL: ForceLongPolling enabled.
> 
> And then I'll be curious to know if your connection issue goes away. Please also provide as much detail as you can about your environment (browser, antivirus software / version, network details, etc.). Thanks!

Hi there,  I tried to use your custom build but seems to me it doesn't work .. tried on last version google chrome . "WEBCHANNEL: ForceLongPolling enabled." in logs , but connections to firestore are in long pending status and then goes down with "Failed to get document because the client is offline." , is anything was changed in the firebase core since your custom build?

**Comment by mikelehen** [comment]
@Ferrasil I looked into this, and it turns out that indeed there was unfortunately a rollback on the backend that broke the forceLongPolling flag used in that test build. :-(  I don't know when the behavior will be re-enabled but I'm looking into it.  Sorry for the confusion.

**Comment by Ferrasil** [comment]
Some updates for everyone affected by this issue, I sent a detailed report to the support and finally they confirmed about the fix in the next patch release which will be in March (version E) ,  details will appear on this page https://support.kaspersky.com/13876

**Comment by pitw** [comment]
@Ferrasil Thanks. Hope that they fix it finally. What did you have to do to convince them of the bug ;) ?

**Comment by Ferrasil** [comment]
@pitw IMHO No hopes they will fix it in that time  lol :) 
Nothing special , I just made a tech support request and did some basics moves like they said : "download it , click record and we will collect data about your PC" => "download it, click record and repeat the bug case so we will see what happen" then they said : "This problem situation # 3056440 will be fixed in the nearest patch E for version 2019. The release of the patch according to preliminary data is scheduled for the end of March."

**Comment by pitw** [comment]
@Ferrasil I already was there ;)

(Watch the date - its from Oct 23 2018 !)

![image](https://user-images.githubusercontent.com/5676857/54030900-5f910d00-41ad-11e9-919b-e05391d8075e.png)

At least they give us the same bug number...

**Comment by Ferrasil** [comment]
@pitw LOL , that's the russian style bug report system , 

> At least they give us the same bug number... 
At least for me now they give us the possible patch release date 🤣

**Comment by kevin192291** [comment]
I am really scared that my end users will be seeing this issue. Basically Kaspersky is preventing them from viewing legitimate content. :(

**Comment by Ferrasil** [comment]
> I am really scared that my end users will be seeing this issue. Basically Kaspersky is preventing them from viewing legitimate content. :(

@kevin192291 lol , I pushed my app to the public almost 2 months ago, and my normal daily work task is to response on user's conflicts in the online support chat , to give them instructions how to add my app to Kaspersky's exclusions, in special cases I should connect to the user's pc with Team Viewer and doing it manually. 
I made a fallback on the  app's "offline" status to catch users and don't let them leave from the website and also I'm really surprised how many users have Kaspersky (my app is for russian community and the antivirus is russian's product , but anyway). So guys sorry for my offtopic here, but this issue hurt my project so hard and I'm losing a lot of customers/money because of that so I can't be silent about that :(

**Comment by mikelehen** [comment]
Just FYI- I believe the workaround I shared on [Jan 4](https://github.com/firebase/firebase-js-sdk/issues/1190#issuecomment-451577257) should be working again if anybody is desperate.  Else, hopefully Kaspersky will fix their fix soon. 🤞

**Comment by ksangular** [comment]
> Just FYI- I believe the workaround I shared on [Jan 4](https://github.com/firebase/firebase-js-sdk/issues/1190#issuecomment-451577257) should be working again if anybody is desperate. Else, hopefully Kaspersky will release their fix soon. 🤞

@mikelehen, How can i add db.settings({forceLongPolling: true}); in a angular test app? If i add its throwing error saying settings method not found.

**Comment by gitEcapital** [comment]
> Just FYI- I believe the workaround I shared on [Jan 4](https://github.com/firebase/firebase-js-sdk/issues/1190#issuecomment-451577257) should be working again if anybody is desperate. Else, hopefully Kaspersky will release their fix soon. 🤞

tried your solution but it didn't work. after i build a with prod flag , 
I got this as console message.

`WEBCHANNEL initializing: 
Object
forceLongPolling: false
httpHeadersOverwriteParam: "$httpHeaders"
httpSessionIdParam: "gsessionid"
initMessageHeaders: {X-Goog-Api-Client: "gl-js/ fire/5.5.9"}
messageUrlParams: {database: "projects/prize-436e3/databases/(default)"}
sendRawJson: true
supportsCrossDomainXhr: true`

**Comment by PetrusStarken** [comment]
> Just FYI- I believe the workaround I shared on [Jan 4](https://github.com/firebase/firebase-js-sdk/issues/1190#issuecomment-451577257) should be working again if anybody is desperate. Else, hopefully Kaspersky will release their fix soon.

Also used your proposed solution, but my app got very slow. I've preferred to not use for now. Anyway, thanks for the work. :)

**Comment by Ferrasil** [comment]
News from battleground ...
I asked Kaspersky support when the patch comes out ,since it's April already , they told me now -  in the end of April :(

**Comment by Ferrasil** [comment]
Hello everyone , as a main newsmaker in this thread , I gladly want to inform that Kaspersky solved the problem and the patch released on 8 April  https://support.kaspersky.com/13876#block1 :
"We have fixed an issue that sometimes caused Kaspersky Anti-Virus to block connections to the Google Cloud Firestore cloud database." 
congratulate everyone who have been touched by this issue and we can close the topic finally 😀

**Comment by nikhilbhalwankar** [comment]
We were also facing the same issue recently with Kaspersky. We took a video call support with Kaspersky and added necessary exceptions for firebase. This resolved the issue.

**Comment by pitw** [comment]
And bug is here again...

> @firebase/firestore: Firestore (5.11.0): Could not reach Cloud Firestore backend. Backend didn't respond within 10 seconds.
> This typically indicates that your device does not have a healthy Internet connection at the moment. The client will operate in offline mode until it is able to successfully connect to the backend.

Happened after upgrading Kaspersky to 19.0.0.1088 (d)

I'm a little bit sceptical if that bug reallly is fixed...

**Comment by anisabboud** [comment]
@pitw hover over the Kaspersky icon.
![image](https://user-images.githubusercontent.com/5672108/60433238-fbd39980-9c0c-11e9-9128-e43e3817d93b.png)
Patch (f) is currently the latest version.
Try restarting your computer to see if Kaspersky got updated.

The bug was fixed in Patch (e).
![image](https://user-images.githubusercontent.com/5672108/60433369-5a007c80-9c0d-11e9-9f62-b07e977f46d9.png)

**Comment by pitw** [comment]
You're right. Thanks.
Kaspersky complained about "not able to update Database" so my genious idea was to "reinstall" with loader from here: https://www.kaspersky.de/downloads/thank-you/internet-security-free-trial

After download, install, restart (1.Time), finish install... Version was still "D" but Database updates worked.
It needed another reboot (2.Time) to get D with Firebase Fix... ;)

**Comment by anisabboud** [comment]
I did the same thing :)
It showed (f) only after reboot.

**Comment by danieldanielecki** [comment]
The problem still exists on the Kaspersky Internet Security newest version...

**Comment by mikelehen** [comment]
@danieldanielecki Can you visit https://debug-my.firebaseapp.com and share your results (e.g. in a gist)?  If this is really happening again, you (or somebody else) will likely need to reach out to Kaspersky as the issue had been solved (as noted in the above release notes and confirmed by several people on this thread).

**Comment by pitw** [comment]
@danieldanielecki As I'm almost paranoid now with Kaspersky + Firebase ;) I tried with several stations which were before concerned from bug.

If you move your mouse over the Kaspersky Icon does it look like this ?
![kaspsfireb](https://user-images.githubusercontent.com/5676857/62353897-f6b78200-b50b-11e9-9dd9-c49e012ae2a7.jpg)

As @anisabboud told above -> **the (f) is important!**

It should work then - double checked on our side.

**Comment by lborgman** [comment]
I think I am affected by this issue as well. Could we please get a way to check (from JavaScript) that this error has occurred? (I can see it in the console, of course.)

That would make it possible to tell users.

**Comment by mikelehen** [comment]
@lborgman Unfortunately, as far as the SDK can tell, the issue is indistinguishable from just being offline or on a slow internet connection...  

FWIW- That message is logged if our connection fails or (in the case of Kaspersky usually) doesn't succeed within 10 seconds.  You could get a similar signal by just doing a `documentReference.get({source: 'server'})` call from your code.  If it fails, then you know the client wasn't able to reach the backend.  But there's no way to know it's because of Kaspersky.

**Comment by lborgman** [comment]
@mikelehen Thanks, that looks nearly like what I want. I guess it is this?

https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/Source

However it would be nice to be able to check if the result I got from the query is from the server. Hm. Maybe it is there and I just have to read the docs?

**Comment by schmidt-sebastian** [comment]
@lborgman Both [QuerySnapshot.getMetadata()](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/QuerySnapshot.html#getMetadata()) and [DocumentSnapshot.getMetadata()](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/DocumentSnapshot.html#getMetadata()) expose this via the [SnapshotMetadata](https://firebase.google.com/docs/reference/android/com/google/firebase/firestore/SnapshotMetadata) object. You can take a look at the `isFromCache()` property, which will be set to false if the snapshot was raised with data from the server.

**Comment by mikelehen** [comment]
FWIW- The corresponding JS SDK references are:
https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentReference.html#get
https://firebase.google.com/docs/reference/js/firebase.firestore.GetOptions.html#source
https://firebase.google.com/docs/reference/js/firebase.firestore.DocumentSnapshot.html#metadata
https://firebase.google.com/docs/reference/js/firebase.firestore.SnapshotMetadata.html#fromcache

**Comment by guillaumevincent** [comment]
If it's fixed It's only fixed on Kaspersky 19+

**Comment by pitw** [comment]
I don't know what exactly from KIS 20 causes the Issue but I'm 99% sure that Kaspersky is responsible for it. Tested on same machine: With Kaspersky 20 installed (Causes Firebase "Network issues"...) and with Kaspersky 20 completly removed (works perfect).

Beside our "beloved" known error "Could not reach Cloud Firestore backend. Backend didn't respond within 10 seconds"... with KIS 20 there's a second error on auth (without any more log infos):

![networkIssue](https://user-images.githubusercontent.com/5676857/65041475-26ed9f80-d957-11e9-9af4-784008f05522.png)


Switching now back to KIS 19 where problem was solved.

**Comment by julianandrade** [comment]
> Same problem here since a week or two. Also Kaspersky...
> When I first load the page the following error message is output (every single time):
> 
> > @firebase/firestore: Firestore (5.5.0): Could not reach Cloud Firestore backend. Backend didn't respond within 10 seconds.
> 
> after 30+ seconds the data from firestore finally loads
> 
> > XHR finished loading: POST "https://firestore.googleapis.com/google.firestore.v1beta1.Firestore/Listen/channel..."
> 
> If I reload the page and navigate to another page, where firestore data is loaded, I immediately get the data back without the error. 😕
> 
> If I pause Kaspersky the error does not occurr.
> 
> ```
> ...
> "@angular/fire": "^5.0.0",
> "firebase": "^5.5.0",
> ...
> ```

I've experienced the exact same issue using firebase, firestore and vuejs.

**Comment by mikelehen** [comment]
@julianandrade I'm not very familiar with Kaspersky products and versioning but based on e.g. https://github.com/firebase/firebase-js-sdk/issues/1190#issuecomment-507228939 I think the original issue was fixed in Kaspersky Antivirus 19.0.0.1088 patch E.  I'm not sure how that relates to the Kaspersky Endpoint Security 11.1 that you are using.

For anybody running into issues with latest versions of Kaspersky software, I'd really encourage you to reach out to Kaspersky and remind them of:
* "We have fixed an issue that sometimes caused Kaspersky Anti-Virus to block connections to the Google Cloud Firestore cloud database." from the [19.0.0.1088 patch E](https://support.kaspersky.com/13876#block1) release notes
* The prior investigation of this issue happened in 3056440.

FYI- For folks that are running into this issue and need a workaround, you can try adding this at the start of your app:

firebase.firestore().settings({
});

**Comment by pitw** [comment]
@julianandrade @mikelehen The version Julian uses is the "business" version. All the settings we talked above can be either fixed in:

- Kaspersky Security Console  (if its hosted in local infrastructure)
- Kaspersky Endpoint Security Cloud (if its the "cloud" variant)

If Julian don't have access to one of these --> theres no way to fix it from "user side" as usually the interesting settings are blocked by admin... 😁

**Comment by julianandrade** [comment]
> @julianandrade I'm not very familiar with Kaspersky products and versioning but based on e.g. [#1190 (comment)](https://github.com/firebase/firebase-js-sdk/issues/1190#issuecomment-507228939) I think the original issue was fixed in Kaspersky Antivirus 19.0.0.1088 patch E. I'm not sure how that relates to the Kaspersky Endpoint Security 11.1 that you are using.
> 
> For anybody running into issues with latest versions of Kaspersky software, I'd really encourage you to reach out to Kaspersky and remind them of:
> 
> * "We have fixed an issue that sometimes caused Kaspersky Anti-Virus to block connections to the Google Cloud Firestore cloud database." from the [19.0.0.1088 patch E](https://support.kaspersky.com/13876#block1) release notes
> * The prior investigation of this issue was associated with bug 3056440 on their end.
> 
> FYI- For folks that are running into this issue and need a workaround, you can try adding this at the start of your app using the current Firestore SDK:
> 
> ```
> firebase.firestore().settings({
>   experimentalForceLongPolling: true
> });
> ```
> 
> This forces the SDK to use a less-efficient long-polling mechanism that we believe works around proxies that buffer traffic indefinitely, like Kaspersky has done in the past.

Yup, that setting worked. Thanks.

**Comment by julianandrade** [comment]
> @julianandrade @mikelehen The version Julian uses is the "business" version. All the settings we talked above can be either fixed in:
> 
> * Kaspersky Security Console  (if its hosted in local infrastructure)
> * Kaspersky Endpoint Security Cloud (if its the "cloud" variant)
> 
> If Julian don't have access to one of these --> theres no way to fix it from "user side" as usually the interesting settings are blocked by admin... 😁

yes, it's a business version and I'll ask our helpdesk team to check the settings. 
Thank you all.

**Comment by DeniCarvalho** [comment]
Guys, one solutions is usage Firestore REST

**Comment by mattiagalati** [comment]
We recently had the same problem with Kaspersky Endpoint version 11.
Using the proposed configuration `experimentalForceLongPolling` seems to have fixed the problem, but we are really thinking about the future of our Javascript applications: even if this is a Kaspersky issue, in more than a year, there was no step forward to completely solve it, also Google in its docs says 
> This setting may be removed in a future release

that said, we cannot plan to develop an application based on this SDK knowing that the workaround could be wiped out and our users could not being able to use our services anymore.

**Comment by wilhuff** [comment]
> This setting may be removed in a future release

The intent of this phrase is that we may remove the option when the need for manually specifying this workaround is gone. Our intent is to fix the SDK to handle this case automatically. Only once that's the case will this option be removed.

**Comment by mattiagalati** [comment]
> > This setting may be removed in a future release
> 
> The intent of this phrase is that we may remove the option when the need for manually specifying this workaround is gone. Our intent is to fix the SDK to handle this case automatically. Only once that's the case will this option be removed.

Thank you @wilhuff , this is reassuring. May I suggest you to change the docs in order to reflect this intent?

**Comment by stei2369** [comment]
[Soundtoys Ultimate Crack VST-Torrents.net](https://vst-torrents.net/soundtoys-ultimate-crack-mac-download/) is a hassle-free software program that enables consumers to create and structure remarkable and revolutionary audio sheets. It is also possible to develop as effortlessly as the dynamic selection of effectiveness. It contains a practical setup with an intuitive person interface.

**Comment by hsubox76** [comment]
Closing as a stale issue. Long polling is still available as an workaround. Please feel free to open a new issue if you are still experiencing this problem.
